add bpftrace scripts to track performance and behavior of SOF functions in the kernel

nklayman opened the pull request:
Signed-off-by: Noah Klayman <noah.klayman@intel.com>

This PR adds some example [bpftrace](https://github.com/iovisor/bpftrace) scripts to track performance and behavior of various SOF functions in the kernel. To run them, you'll need to [install bpftrace](https://github.com/iovisor/bpftrace/blob/master/INSTALL.md), then run the script with `bpftrace [path to script]`. If you want to use the hda_irq scripts, you'll need to be running [this kernel pr](https://github.com/thesofproject/linux/pull/3482).

Example output from `suspend_resume_time.bt`:
```shell
❯ bpftrace ./kernel_tracing/bpftrace_scripts/suspend_resume_time.bt
Attaching 5 probes...
kprobe:sof_resume (runtime) started
kretprobe:sof_resume (runtime) finished, took 132135788 nsecs
kprobe:sof_suspend (system) started
kretprobe:sof_suspend (system) finished, took 738362 nsecs
kprobe:sof_resume (system) started
kretprobe:sof_resume (system) finished, took 197517950 nsecs
kprobe:sof_suspend (runtime) started
kretprobe:sof_suspend (runtime) finished, took 32807764 nsecs
kprobe:sof_resume (runtime) started
kretprobe:sof_resume (system) finished, took 187189967 nsecs
kprobe:sof_resume (system) started
kretprobe:sof_resume (system) finished, took 180462467 nsecs
kprobe:sof_suspend (runtime) started
kretprobe:sof_suspend (runtime) finished, took 32866837 nsecs
kprobe:sof_resume (runtime) started
kprobe:sof_resume (system) started
kretprobe:sof_resume (system) finished, took 188085072 nsecs
kprobe:sof_resume (runtime) started
kretprobe:sof_resume (system) finished, took 188318490 nsecs
kprobe:sof_suspend (runtime) started
kretprobe:sof_suspend (runtime) finished, took 32849861 nsecs
^C


@usecs[kretprobe:sof_resume, runtime]: 
[128K, 256K)           5 |@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@|

@usecs[kretprobe:sof_resume, system]: 
[128K, 256K)           5 |@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@|

@usecs[kretprobe:sof_suspend, runtime]: 
[16K, 32K)             2 |@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@                  |
[32K, 64K)             3 |@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@|

@usecs[kretprobe:sof_suspend, system]: 
[256, 512)             1 |@@@@@@@@@@@@@                                       |
[512, 1K)              4 |@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@|
```

Additionally, this PR includes a Python script to collect metrics and run conditions on them, see `kernel_tracing/metric_evaluation`. It's intended to be used on the CI machines to prevent performance regressions.

plbossart commented on tracing/suspend_resume_time.bt:
```diff
@@ -0,0 +1,24 @@
+// This tracks the time it takes for runtime and system suspend/resume
```
// SPDX-License-Identifier: BSD-3-Clause
// Copyright(c) 2022 Intel Corporation. All rights reserved.

nklayman replied:
None of the files in sof-test had this, are you sure it's necessary?

plbossart replied:
I prefer to avoid any ambiguity related to license/copyright.

plbossart commented on tracing/suspend_resume_time.bt:
```diff
@@ -0,0 +1,24 @@
+// This tracks the time it takes for runtime and system suspend/resume
```
commit message: you need to have some sort of prefix, then the action, e.g.

tracing: add suspend and resume time tracking with bptrace

Also add some of the explanations in the commit message itself, not the PR.

nklayman replied:
Good point, just updated.

plbossart reviewed (changes requested):
looks quite promising @nklayman, minor nit-picks to improve the commit below.

plbossart commented:
added the 'kernel maintainer' team for feedback.

marc-hb reviewed (commented):
Can you rename the directory to something more specific, maybe `sof-test/linux/tracing/` or `sof-test/ktracing`? We have a lot of "tracing" already and `sof-test/tracing/` sounds definitely like firmware tracing.

Can you add a README with some pointers, context/purpose/intro/etc. or if too short for a README then some comments in this new `.bt` file? For now it's very... "raw".

nklayman commented:
Just added a readme and updated the directory structure to make it more clear what this is.

plbossart commented:
> Just added a readme and updated the directory structure to make it more clear what this is.

it's better to squash all the commits together and do a git push --force to have a single commit, no?

marc-hb commented:
> it's better to squash all the commits together and do a git push --force to have a single commit, no?

@nklayman has just used Github the way it was designed to be used, not the way we use it.

Github tolerates but does not really support rewriting history and force pushes. Github is optimized for a single commit per pull request + fixup commits just like @nklayman did. For instance try the "View changes" button above: this time it works! Same for many other Github features.

More info at https://github.com/zephyrproject-rtos/zephyr/issues/39194

Note Github has an option to squash all commits right before merging. Not sure what is the final commit message then.

nklayman commented:
I prefer to push individually and then squash into one commit before merging, either manually or through the Github UI. Otherwise, it's a lot harder to tell what's changed and see what code review comments refer to. I can do the single commit method if you'd prefer, but as long as we squash before merging the end result is the same, and the process to get there is a bit better.

marc-hb commented:
_As long as the goal is to have a single commit in the end_, I totally agree the process is much better and the end result is the same. It is the way Github is designed to be used after all!

The problem is purely about **consistency**: as Github PRs do not support reviewing multiple commits "natively", this means learning how to use Github TWICE, in two different ways. One for reviewing a single commit (in the end), and a second workflow for reviewing multiple commits (in the end).

I don't mind at all but I've been the "local git guru" in many teams for many years so I'm not representative. I've seen many, many people struggling with ONE way to use git or Github.

marc-hb commented on kernel_tracing/metric_evaluation/test.py:
```diff
@@ -0,0 +1,112 @@
+    lines = self.process.stdout.read().split("\n")
+    vars = {}
+    for line in lines:
+      if re.match("@.*: .*", line):
```
Regular expressions are "write-only" code so please provide one example as a comment.

marc-hb commented on kernel_tracing/metric_evaluation/spec.json:
```diff
@@ -0,0 +1,26 @@
+        "@avg[kretprobe:sof_resume, runtime] < 150000",
+        "@avg[kretprobe:sof_resume, system] < 200000",
+        "@avg[kretprobe:sof_suspend, runtime] < 50000",
+        "@avg[kretprobe:sof_suspend, system] < 1000"
```
Does `@avg` mean "average"? Averages hide intermittent bugs, they're not very interesting. 

https://bravenewgeek.com/everything-you-know-about-latency-is-wrong/

nklayman replied:
Yes, it means average. However, it's purely a demo (I just made the numbers up), so it doesn't matter much. I figured it'd be better to let those more familiar with SOF figure out the real metrics/conditions to use. I assume we'll end up with some sort of script to generate test cases with device-specific conditions and tests anyways. I've renamed the spec file to `sample_spec.json` to make it more clear that it's just a demo.

marc-hb commented on kernel_tracing/metric_evaluation/test.py:
```diff
@@ -0,0 +1,112 @@
+  # Execute shell commands
+  for cmd in shell_commands:
+    print_color(colors.BLUE, f"$ {cmd}")
+    subprocess.call(cmd, cwd=working_dir, shell=True)
```
https://docs.python.org/3/library/subprocess.html#using-the-subprocess-module
> The recommended approach to invoking subprocesses is to use the [run()](https://docs.python.org/3/library/subprocess.html#subprocess.run) function for all use cases it can handle. For more advanced use cases, the underlying [Popen](https://docs.python.org/3/library/subprocess.html#subprocess.Popen) interface can be used directly.

https://docs.python.org/3/library/subprocess.html#older-high-level-api
> Prior to Python 3.5, these three functions comprised the high level API to subprocess. You can now use [run()](https://docs.python.org/3/library/subprocess.html#subprocess.run) in many cases, but lots of existing code calls these functions.
> `subprocess.call(args, *, ...`

marc-hb commented on kernel_tracing/metric_evaluation/test.py:
```diff
@@ -0,0 +1,112 @@
+    condition = condition.replace(key, val)
+  return eval(condition)
+
+def run_test(test_case, working_dir):
```
`check_condition`, `run_test`, `print_color`,... I'm not familiar with `pytest` but a lot of this script looks like re-implementing a test framework. We unfortunately already made that mistake in this sof-test repo + in some internal repos and I'd hate doing it again.

tl;dr: did you take a quick look at `pytest`?

nklayman replied:
The reason I did this all manually is so that the tests can be defined in a JSON file, which makes it a lot simpler to write/manage tests. While I could convert this script to a library and have the tests written in Python, that would reduce the simplicity of writing tests. JSON is also easy to manipulate with a program, so it'd be much easier to design a script to generate test suites that are customized per-device. If you're just as happy to write the test cases in Python, then I'll make this a library that works with `pytest`. Otherwise, I'll look into using `pytest` with the current JSON-based test definitions.

marc-hb replied:
> Otherwise, I'll look into using pytest with the current JSON-based test definitions.

Yes.

I'm not very familiar with pytest but I don't see why it would stop you from configuring tests in JSON. I think it's pretty flexible.

nklayman replied:
The issue is that pytest organizes test cases into functions. Therefore, I'd need to create a function for each test case in the JSON which isn't really possible. If each case didn't have it's own function, the error output would be pretty useless as you wouldn't even know which test case is failing. If pytest has some way to dynamically create tests, then it would work great. I'll do some more investigating to see if there is a method for this.

marc-hb replied:
I think pytest supports test parameters. So hopefully you can have a single function for all the tests. Random example (found very quickly, there are probably better ones)
https://docs.pytest.org/en/6.2.x/example/parametrize.html

```
  <Function test_timedistance_v2[20011212-20011211-expected0]>
  <Function test_timedistance_v2[20011211-20011212-expected1]>
```

nklayman replied:
After many hours, I was able to wrangle pytest so that it works with the `spec.json`. Getting the output formatting right took forever, but I'm really happy with the end result:
![image](https://user-images.githubusercontent.com/12448243/161403862-7656e2fe-059f-40df-b120-dd92662d142b.png)

Give it a try and let me know if you have any more suggestions.

marc-hb commented on kernel_tracing/metric_evaluation/test.py:
```diff
@@ -0,0 +1,112 @@
+      print_color(colors.GREEN, "Test case passed")
+
+  if has_failed:
+    sys.exit(1)
```
Another re-implementation of a basic test framework feature.

marc-hb reviewed (changes requested):
Don't forget to run `pylint` now and then. It's not gospel but it's pretty good.

marc-hb commented on kernel_tracing/metric_evaluation/test.py:
You can dynamically generate one test instance per condition at run-time thanks to `pytest_generate_tests()`, example at https://stackoverflow.com/questions/49076378/pytest-generate-tests-based-on-arguments

This will save a lot of the code in `check_conditions()` (now `check_condition()`, singular) and produce a much more readable output.

nklayman replied:
I switched to doing it entirely through parameterization (haven't pushed yet), and I think it makes the error output worse. Now, the report lists each condition instead of just the cases, as each condition is an individual parameterized test.

![image](https://user-images.githubusercontent.com/12448243/162331469-49cf4265-b786-4a01-b6b7-dec328595f84.png)

Do you want me to go forward with this or revert back to the original setup?

marc-hb replied:
> Now, the report lists each condition instead of just the cases, as each condition is an individual parameterized test.

That was indeed my request: moving PASS / FAIL statuses to pytest instead of manually handling and presenting test results. What's wrong with it?

pytest has the ability to output test results in various formats including the junitXML standard for integration with CI. Handling failed conditions manually means it would not be possible for CI to show which conditions have failed and which ones have passed.

https://docs.pytest.org/en/7.1.x/how-to/output.html
https://docs.pytest.org/en/6.2.x/usage.html#creating-junitxml-format-files

marc-hb commented on kernel_tracing/metric_evaluation/test.py:
Starting and stopping the BPF trace is not part of the test but part of the test "fixture" https://docs.pytest.org/en/7.1.x/how-to/fixtures.html#fixtures-can-be-requested-more-than-once-per-test-return-values-are-cached

marc-hb replied:
More relevant link sorry:
https://docs.pytest.org/en/7.1.x/how-to/fixtures.html#scope-sharing-fixtures-across-classes-modules-packages-or-session

nklayman commented on kernel_tracing/metric_evaluation/test.py:
```diff
@@ -0,0 +1,113 @@
+            # Check if line is outputting a variable from the script
+            # bpftrace output format is `@name[key1, key2, ...] = val` (keys are optional)
+            if re.match("@.*: .*", line):
+                key, val = line.split(": ", 1)
```
`name` would be the name of the map, which can be indexed by any number of keys. ie `@max_time[ipc, stream] = 500` would be the map `max_time` indexed by two keys, in this case the value `500` is tied to keys `ipc` and `stream`. It's essentially a flattened nested map.

nklayman commented:
@marc-hb all your comments should be addressed in the latest update. I switched it over to the one-condition-per-test format that you prefer, and applied the small changes. @plbossart I also squashed the commits together into two as you requested, one for the bpftrace scripts themselves and one for the metric evaluation part.

Marking this "ready for review" now as it's at a good state for others to look at. The metric evaluation script will most likely need some changes to make it work in the CI, but that's probably a better task for the CI devs rather than myself.

marc-hb commented on kernel_tracing/metric_evaluation/test.py:
```diff
@@ -0,0 +1,54 @@
+
+def test(case):
+    """ This is the test function that is called by pytest """
+    (_, condition, bpftrace_vars) = case
```
```suggestion
    (_ignored_something, condition, bpftrace_vars) = case
```
From an "API" perspective it's useful to know what gets discarded.

aiChaoSONG commented on kernel_tracing/metric_evaluation/conftest.py:
```diff
@@ -0,0 +1,89 @@
+
+def pytest_generate_tests(metafunc):
+    """ Generate parameter set for each test case in json file """
+    if "case" in metafunc.fixturenames:
```
`fixturenames`, I think this is a typo

aiChaoSONG commented on kernel_tracing/metric_evaluation/test.py:
```diff
@@ -0,0 +1,54 @@
+def test(case):
+    """ This is the test function that is called by pytest """
+    (_, condition, bpftrace_vars) = case
+    conditionWithVars = condition
```
let's use snake_case

nklayman commented on kernel_tracing/metric_evaluation/test.py:
```diff
@@ -0,0 +1,54 @@
+If any condition fails this script exits with code 1, otherwise it exits with code 0.
+Shell commands are run in the directory of the spec file.
+The bpftrace script path is relative to the spec file as well.
+See sample_spec.json for an example.
```
I ended up fixing this because it was easier than updating the help message.

nklayman commented on kernel_tracing/metric_evaluation/sample_spec.json:
```diff
@@ -0,0 +1,26 @@
+    {
+      "name": "IPC TX Time",
+      "script": "../bpftrace_scripts/ipc_tx_time.bt",
+      "shell": [
```
I think using "shell" as the name still makes sense here, but happy to change it if you disagree.

nklayman commented:
@marc-hb Thanks for the review! Almost all your comments should be addressed now. The no-dependency test option is a bit tricky because the script can't really work without bpftrace. If you're okay with having that a dependency I could make a very simple case, otherwise I could try aliasing `bpftrace` to echo out some sample output so there's truly no dependencies. Let me know which one you prefer.

marc-hb commented:
> The no-dependency test option is a bit tricky because the script can't really work without bpftrace

The dependency on bpftrace user space is not a problem because it's as simple as `apt install bpftrace`. The annoying part is the kernel part. Reconfiguring and rebooting a kernel before submitting a fix for a minor python warning is not convenient. So I'm just wondering if it would be possible to have some kind of "bpftrace no-op" example / test case for this = anything that does not require a new kernel. If bpftrace has some hard kernel dependencies that are typically not available in stock kernels provided by distributions then too bad forget it. Maybe we could add some sort of "mock bpftrace" later.

(I've not looked at the new version yet)

marc-hb reviewed (approved):
I have not tried to run it again but I find the entire code very clear and easy to read and none of this is running in production yet so bug fixes and incremental improvements can be done later: I recommend we merge this to make it easier for others to try.

@plbossart could you have another look, especially the `.bt` files that I'm not familiar with?

marc-hb reviewed (dismissed):
Some review comments from a generous `pytest` expert friend who prefers to stay anonymous and not involved:
>  i don't know bpftrace but it looks a little bit like you might want to be using a parameterized fixture to run the test cases instead.
> simple example usage: https://github.com/zephyrproject-rtos/west/blob/a343b7a7982355d0e5521ddcc69f700940d89e50/tests/test_manifest.py#L2885
> with a parameterized fixture, you'll get finer-grained reporting and can customize pytest's notions of what each test is. as-is, it looks like you'll get a global pass/fail on a single function

I think we have parametrization already but maybe `@pytest.mark.parametrize` is more "pythonic", flexible and generally better way to do it. Again this can be a future improvement after merge.

marc-hb commented:
> I think we have parametrization already but maybe @pytest.mark.parametrize is more "pythonic", flexible and generally better way to do it

I just remembered reading somewhere complaining that `@pytest.mark.parametrize` is static and can't be used after reading a config file. Vague and second-hand information, sorry for any noise.

nklayman commented:
@marc-hb latest push includes a `no_dependencies.json` that should work with just bpftrace installed. Also fixes edge cases where the spec file arg doesn't start with a `./` and when the bpftrace script doesn't exit in time.

marc-hb commented:
Nice try but:
```
collecting ... modprobe: FATAL: Module kheaders not found in directory /lib/modules/9.9.9-rc9softest
$ ['sleep', '1']
BPF tracing finished
```

Also, not running as root still hangs:
```
collecting ... ERROR: bpftrace currently only supports running as the root user.
  ... hangs ...
```

```
$ ps f

    987 pts/0    Ss     0:00 -bash
   1161 pts/0    R+     0:39  \_ python3 ./kernel_tracing/metric_evaluation/test.py kernel_tracing/metric_evaluation/no_dependencies.json
   1163 pts/0    Z+     0:00      \_ [bpftrace] <defunct>
```

BTW only bpftrace should run as root, should be started as `sudo bpftrace`.

Again all this can be fixed later.

plbossart commented on kernel_tracing/bpftrace_scripts/hda_irq_source_counts.bt:
```diff
@@ -0,0 +1,12 @@
+// SPDX-License-Identifier: BSD-3-Clause
```
typo in commit message: timestamp.

plbossart commented on kernel_tracing/metric_evaluation/simple_bpftrace.bt:
```diff
@@ -0,0 +1,7 @@
+
+BEGIN {
+    @val = 42;
+}
```
I don't get what this is for and what '42' might possibly represent - other that "the answer is always 42"

nklayman replied:
The answer is always 42 for this script. The purpose is to be able to run the python framework without having any of the kernel dependencies for bpf installed.

plbossart replied:
Ah ok, so now I get it.

maybe rename to '@test_check' or something. 'val' threw me off, I didn't understand what this was.

plbossart commented on kernel_tracing/metric_evaluation/no_dependencies.json:
```diff
@@ -0,0 +1,12 @@
+        ]
+      }
+    ]
+  }
```
no idea what this does?

nklayman replied:
It declares a test case for the python script to run. The help message in the python script explains how to execute it. I could add additional documentation if you think that makes sense.

marc-hb replied:
I don't remember the specific but for a very first example / demo I would generally err on the "too much documentation" side. Don't get me wrong: I mean "too much" as "rephrasing the same thing twice", NOT as "digressions about more advanced stuff".

nklayman replied:
I added a readme in the metric_evaluation folder that explains what this all does.

plbossart reviewed (commented):
the code split and what file is used for what isn't very clear to me @nklayman

nklayman commented on kernel_tracing/bpftrace_scripts/suspend_resume_time.bt:
```diff
@@ -19,6 +19,7 @@ kretprobe:sof_resume, kretprobe:sof_suspend {
   // When the function returns, we use the saved timestamp to determine execution time
   printf("%s (%s) finished, took %d nsecs\n", probe, @type, nsecs - @start);
   @usecs[probe, @type] = hist((nsecs - @start) / 1000);
+  @avg[probe, @type] = avg((nsecs - @start) / 1000);
```
I added it as part of that commit since the Python script can't operate on the histogram value, so it needs another value to use for the threshold checking. I think this will end up all squashed together anyways, but I can move it to the previous commit if you'd prefer.

plbossart replied:
I think it'd be clearer if all bpf scripts were moved to the previous patch

plbossart commented on kernel_tracing/metric_evaluation/readme.md:
```diff
@@ -0,0 +1,34 @@
+        "list of conditions that will be evaluated against bpftrace output",
+        "these are python expressions",
+        "bpftrace vars can be referenced with @{var name}, ie @avg",
+        "maps are referenced by @{var name}[{key name(s)}], ie @avg[ipc_time]"
```
so in the example above, that does '@val' refer to?

nklayman replied:
That refers to the static val = 42 assignment in the no_dependencies script. I'll be renaming it to test_value so it's more clear.

plbossart reviewed (commented):
mostly ok @nklayman, just tripped on the 'val' variable that could really mean anything.

@marc-hb you good with this version?

nklayman commented:
BPF script changes have been moved to the first patch and `var` has been renamed. @marc-hb @plbossart everything look good now?

plbossart reviewed (approved):
LGTM, I'll let @marc-hb comment further on the Python parts. Thanks @nklayman !

marc-hb reviewed (approved):
Great start, very clear and readable. None of this is used in production yet so bugs can always be fixed later.
